### PR TITLE
Add Pydantic Json field support

### PR DESCRIPTION
**Pull request by hramezani**
Fixes https://github.com/pydantic/pydantic-settings/issues/78

Selected Reviewer: @samuelcolvin

### CODE REVIEW

**Comment by hramezani** [comment]
please review

File: pydantic_settings/sources.py [hunk 1]
```diff
@@ -602,7 +602,9 @@ def read_env_file(
         return file_vars
 
 
-def _annotation_is_complex(annotation: type[Any] | None) -> bool:
+def _annotation_is_complex(annotation: type[Any] | None, metadata: list[Any]) -> bool:
+    if any(str(md) == 'Json' for md in metadata):
```
Comment (hramezani) [inline pydantic_settings/sources.py:606]: Not sure I am doing the right way to check if the field is a `Json` field.

Comment (dmontagu) [reply pydantic_settings/sources.py:606]: Yeah this seems to be right:
```suggestion
    if any(isinstance(md, Json) for md in metadata):
```

(Also need to add `Json` to the set of imports from `pydantic` at the top)

Comment (hramezani) [reply pydantic_settings/sources.py]: Thanks for your suggestion. Done

File: pydantic_settings/sources.py [hunk 2]
```diff
@@ -602,7 +602,9 @@ def read_env_file(
         return file_vars
 
 
-def _annotation_is_complex(annotation: type[Any] | None) -> bool:
+def _annotation_is_complex(annotation: type[Any] | None, metadata: list[Any]) -> bool:
+    if any(isinstance(md, Json) for md in metadata):  # type: ignore[misc]
```
Comment (dmontagu) [inline pydantic_settings/sources.py:606]:
```suggestion
    if any(isinstance(md, Json) for md in metadata):
```
Was the type ignore here necessary? I didn't think I needed it to run mypy with no errors but maybe I missed something. Just making sure

Comment (hramezani) [reply pydantic_settings/sources.py:606]: It failed without type ignore
Here is the log: https://github.com/pydantic/pydantic-settings/actions/runs/5357133312/jobs/9717574129

**Review by dmontagu — APPROVED** [review]

**Pull request merged**